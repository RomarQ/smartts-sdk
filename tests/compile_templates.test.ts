import fs from 'fs';
import SmartML from '../src/smartml';

jest.setTimeout(50000);

const verifyMichelsonOutput = (contract: string) => {
    const michelson = SmartML.compileContract(contract);

    // No errors expected
    // Check snapshot
    expect(michelson).toMatchSnapshot();
    expect(JSON.stringify(michelson).includes('ERROR')).toBeFalsy();
};

async function runTests() {
    const files = fs.readdirSync('./tests/templates', { encoding: 'utf-8' });
    files.forEach(async (fileName) => {
        test(fileName, async () => {
            try {
                const contract = (await import(`./templates/${fileName}`)).default.toString();
                expect(contract).toMatchSnapshot();
                verifyMichelsonOutput(contract);
            } catch (e) {
                console.debug(e);
            }
        });
    });
}

describe('Test contract templates', () => {
    runTests();
});
