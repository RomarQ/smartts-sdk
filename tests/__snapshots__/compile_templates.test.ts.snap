// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test contract templates FA2.ts 1`] = `
"
        (
            template_id (static_id 0 (\\"FA2.ts\\" 42))
            storage (record (\\"FA2.ts\\" 62) (config (record (\\"FA2.ts\\" 63) (administrator (literal (address tz1) (\\"FA2.ts\\" 64))) (paused (literal (bool False) (\\"FA2.ts\\" 65))))) (assets (record (\\"FA2.ts\\" 67) (ledger (big_map (\\"FA2.ts\\" 68) )) (operators (big_map (\\"FA2.ts\\" 69) )) (token_metadata (big_map (\\"FA2.ts\\" 70) )) (token_total_supply (big_map (\\"FA2.ts\\" 71) )))) (metadata (big_map (\\"FA2.ts\\" 73) )))
            storage_type ((record ((config (record ((administrator \\"address\\") (paused \\"bool\\")) (Some ((\\"administrator\\") (\\"paused\\"))))) (assets (record ((ledger (bigmap (tuple \\"address\\" \\"nat\\") (record ((balance \\"nat\\")) (Some Right)))) (operators (bigmap (record ((operator \\"address\\") (owner \\"address\\") (token_id \\"nat\\")) (Some Right)) \\"unit\\")) (token_metadata (bigmap \\"nat\\" (record ((token_id \\"nat\\") (token_info (map \\"string\\" \\"bytes\\"))) (Some Right)))) (token_total_supply (bigmap \\"nat\\" \\"nat\\"))) (Some ((((\\"ledger\\") (\\"operators\\")) (\\"token_metadata\\")) (\\"token_total_supply\\"))))) (metadata (bigmap \\"string\\" \\"bytes\\"))) (Some Right)))
            messages ((transfer True False False False (\\"FA2.ts\\" 76) ()) (update_operators True False False False (\\"FA2.ts\\" 77) ()) (mint True False False False (\\"FA2.ts\\" 79) ((verify (eq (sender) (attr (attr (data) \\"config\\" (\\"expression.ts\\" 26)) \\"administrator\\" (\\"expression.ts\\" 26)) (\\"FA2.ts\\" 79)) (literal (string \\"FA2__Not_Admin\\") (\\"FA2.ts\\" 79)) (\\"FA2.ts\\" 79)))) (pause True False False True (\\"FA2.ts\\" 85) ((set_type (params (\\"FA2.ts\\" 85)) \\"bool\\" (\\"FA2.ts\\" 85)) (verify (eq (sender) (attr (attr (data) \\"config\\" (\\"expression.ts\\" 26)) \\"administrator\\" (\\"expression.ts\\" 26)) (\\"FA2.ts\\" 85)) (literal (string \\"FA2__Not_Admin\\") (\\"FA2.ts\\" 85)) (\\"FA2.ts\\" 85)) (set (attr (attr (data) \\"config\\" (\\"expression.ts\\" 26)) \\"paused\\" (\\"expression.ts\\" 26)) (params (\\"FA2.ts\\" 85)) (\\"FA2.ts\\" 85)))) (set_admin True False False True (\\"FA2.ts\\" 93) ((set_type (params (\\"FA2.ts\\" 93)) \\"address\\" (\\"FA2.ts\\" 93)) (verify (eq (sender) (attr (attr (data) \\"config\\" (\\"expression.ts\\" 26)) \\"administrator\\" (\\"expression.ts\\" 26)) (\\"FA2.ts\\" 93)) (literal (string \\"FA2__Not_Admin\\") (\\"FA2.ts\\" 93)) (\\"FA2.ts\\" 93)) (set (attr (attr (data) \\"config\\" (\\"FA2.ts\\" 93)) \\"administrator\\" (\\"FA2.ts\\" 93)) (params (\\"FA2.ts\\" 93)) (\\"FA2.ts\\" 93)))) (update_metadata True False False True (\\"FA2.ts\\" 100) ((set_type (params (\\"FA2.ts\\" 100)) (map \\"string\\" \\"bytes\\") (\\"FA2.ts\\" 100)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"FA2.ts\\" 42))
        )
        "
`;

exports[`Test contract templates FA2.ts 2`] = `
Array [
  Object {
    "args": Array [
      Object {
        "args": Array [
          Object {
            "annots": Array [
              "%config",
            ],
            "args": Array [
              Object {
                "annots": Array [
                  "%administrator",
                ],
                "prim": "address",
              },
              Object {
                "annots": Array [
                  "%paused",
                ],
                "prim": "bool",
              },
            ],
            "prim": "pair",
          },
          Object {
            "args": Array [
              Object {
                "annots": Array [
                  "%assets",
                ],
                "args": Array [
                  Object {
                    "args": Array [
                      Object {
                        "args": Array [
                          Object {
                            "annots": Array [
                              "%ledger",
                            ],
                            "args": Array [
                              Object {
                                "args": Array [
                                  Object {
                                    "prim": "address",
                                  },
                                  Object {
                                    "prim": "nat",
                                  },
                                ],
                                "prim": "pair",
                              },
                              Object {
                                "prim": "nat",
                              },
                            ],
                            "prim": "big_map",
                          },
                          Object {
                            "annots": Array [
                              "%operators",
                            ],
                            "args": Array [
                              Object {
                                "args": Array [
                                  Object {
                                    "annots": Array [
                                      "%operator",
                                    ],
                                    "prim": "address",
                                  },
                                  Object {
                                    "args": Array [
                                      Object {
                                        "annots": Array [
                                          "%owner",
                                        ],
                                        "prim": "address",
                                      },
                                      Object {
                                        "annots": Array [
                                          "%token_id",
                                        ],
                                        "prim": "nat",
                                      },
                                    ],
                                    "prim": "pair",
                                  },
                                ],
                                "prim": "pair",
                              },
                              Object {
                                "prim": "unit",
                              },
                            ],
                            "prim": "big_map",
                          },
                        ],
                        "prim": "pair",
                      },
                      Object {
                        "annots": Array [
                          "%token_metadata",
                        ],
                        "args": Array [
                          Object {
                            "prim": "nat",
                          },
                          Object {
                            "args": Array [
                              Object {
                                "annots": Array [
                                  "%token_id",
                                ],
                                "prim": "nat",
                              },
                              Object {
                                "annots": Array [
                                  "%token_info",
                                ],
                                "args": Array [
                                  Object {
                                    "prim": "string",
                                  },
                                  Object {
                                    "prim": "bytes",
                                  },
                                ],
                                "prim": "map",
                              },
                            ],
                            "prim": "pair",
                          },
                        ],
                        "prim": "big_map",
                      },
                    ],
                    "prim": "pair",
                  },
                  Object {
                    "annots": Array [
                      "%token_total_supply",
                    ],
                    "args": Array [
                      Object {
                        "prim": "nat",
                      },
                      Object {
                        "prim": "nat",
                      },
                    ],
                    "prim": "big_map",
                  },
                ],
                "prim": "pair",
              },
              Object {
                "annots": Array [
                  "%metadata",
                ],
                "args": Array [
                  Object {
                    "prim": "string",
                  },
                  Object {
                    "prim": "bytes",
                  },
                ],
                "prim": "big_map",
              },
            ],
            "prim": "pair",
          },
        ],
        "prim": "pair",
      },
    ],
    "prim": "storage",
  },
  Object {
    "args": Array [
      Object {
        "args": Array [
          Object {
            "args": Array [
              Object {
                "annots": Array [
                  "%mint",
                ],
                "prim": "unit",
              },
              Object {
                "args": Array [
                  Object {
                    "annots": Array [
                      "%pause",
                    ],
                    "prim": "bool",
                  },
                  Object {
                    "annots": Array [
                      "%set_admin",
                    ],
                    "prim": "address",
                  },
                ],
                "prim": "or",
              },
            ],
            "prim": "or",
          },
          Object {
            "args": Array [
              Object {
                "annots": Array [
                  "%transfer",
                ],
                "prim": "unit",
              },
              Object {
                "args": Array [
                  Object {
                    "annots": Array [
                      "%update_metadata",
                    ],
                    "args": Array [
                      Object {
                        "prim": "string",
                      },
                      Object {
                        "prim": "bytes",
                      },
                    ],
                    "prim": "map",
                  },
                  Object {
                    "annots": Array [
                      "%update_operators",
                    ],
                    "prim": "unit",
                  },
                ],
                "prim": "or",
              },
            ],
            "prim": "or",
          },
        ],
        "prim": "or",
      },
    ],
    "prim": "parameter",
  },
  Object {
    "args": Array [
      Array [
        Object {
          "prim": "UNPAIR",
        },
        Object {
          "args": Array [
            Array [
              Object {
                "args": Array [
                  Array [
                    Object {
                      "prim": "DROP",
                    },
                    Object {
                      "prim": "DUP",
                    },
                    Object {
                      "prim": "CAR",
                    },
                    Object {
                      "prim": "CAR",
                    },
                    Object {
                      "prim": "SENDER",
                    },
                    Object {
                      "prim": "COMPARE",
                    },
                    Object {
                      "prim": "EQ",
                    },
                    Object {
                      "args": Array [
                        Array [],
                        Array [
                          Object {
                            "args": Array [
                              Object {
                                "prim": "string",
                              },
                              Object {
                                "string": "FA2__Not_Admin",
                              },
                            ],
                            "prim": "PUSH",
                          },
                          Object {
                            "prim": "FAILWITH",
                          },
                        ],
                      ],
                      "prim": "IF",
                    },
                  ],
                  Array [
                    Object {
                      "args": Array [
                        Array [
                          Object {
                            "prim": "SWAP",
                          },
                          Object {
                            "prim": "DUP",
                          },
                          Object {
                            "args": Array [
                              Object {
                                "int": "2",
                              },
                            ],
                            "prim": "DUG",
                          },
                          Object {
                            "prim": "CAR",
                          },
                          Object {
                            "prim": "CAR",
                          },
                          Object {
                            "prim": "SENDER",
                          },
                          Object {
                            "prim": "COMPARE",
                          },
                          Object {
                            "prim": "EQ",
                          },
                          Object {
                            "args": Array [
                              Array [],
                              Array [
                                Object {
                                  "args": Array [
                                    Object {
                                      "prim": "string",
                                    },
                                    Object {
                                      "string": "FA2__Not_Admin",
                                    },
                                  ],
                                  "prim": "PUSH",
                                },
                                Object {
                                  "prim": "FAILWITH",
                                },
                              ],
                            ],
                            "prim": "IF",
                          },
                          Object {
                            "prim": "SWAP",
                          },
                          Object {
                            "prim": "DUP",
                          },
                          Object {
                            "prim": "CAR",
                          },
                          Object {
                            "args": Array [
                              Object {
                                "int": "2",
                              },
                            ],
                            "prim": "DIG",
                          },
                          Object {
                            "args": Array [
                              Object {
                                "int": "2",
                              },
                            ],
                            "prim": "UPDATE",
                          },
                          Object {
                            "args": Array [
                              Object {
                                "int": "1",
                              },
                            ],
                            "prim": "UPDATE",
                          },
                        ],
                        Array [
                          Object {
                            "prim": "SWAP",
                          },
                          Object {
                            "prim": "DUP",
                          },
                          Object {
                            "args": Array [
                              Object {
                                "int": "2",
                              },
                            ],
                            "prim": "DUG",
                          },
                          Object {
                            "prim": "CAR",
                          },
                          Object {
                            "prim": "CAR",
                          },
                          Object {
                            "prim": "SENDER",
                          },
                          Object {
                            "prim": "COMPARE",
                          },
                          Object {
                            "prim": "EQ",
                          },
                          Object {
                            "args": Array [
                              Array [],
                              Array [
                                Object {
                                  "args": Array [
                                    Object {
                                      "prim": "string",
                                    },
                                    Object {
                                      "string": "FA2__Not_Admin",
                                    },
                                  ],
                                  "prim": "PUSH",
                                },
                                Object {
                                  "prim": "FAILWITH",
                                },
                              ],
                            ],
                            "prim": "IF",
                          },
                          Object {
                            "prim": "SWAP",
                          },
                          Object {
                            "prim": "DUP",
                          },
                          Object {
                            "prim": "CAR",
                          },
                          Object {
                            "args": Array [
                              Object {
                                "int": "2",
                              },
                            ],
                            "prim": "DIG",
                          },
                          Object {
                            "args": Array [
                              Object {
                                "int": "1",
                              },
                            ],
                            "prim": "UPDATE",
                          },
                          Object {
                            "args": Array [
                              Object {
                                "int": "1",
                              },
                            ],
                            "prim": "UPDATE",
                          },
                        ],
                      ],
                      "prim": "IF_LEFT",
                    },
                  ],
                ],
                "prim": "IF_LEFT",
              },
            ],
            Array [
              Object {
                "args": Array [
                  Array [
                    Object {
                      "prim": "DROP",
                    },
                  ],
                  Array [
                    Object {
                      "args": Array [
                        Array [
                          Object {
                            "prim": "DROP",
                          },
                        ],
                        Array [
                          Object {
                            "prim": "DROP",
                          },
                        ],
                      ],
                      "prim": "IF_LEFT",
                    },
                  ],
                ],
                "prim": "IF_LEFT",
              },
            ],
          ],
          "prim": "IF_LEFT",
        },
        Object {
          "args": Array [
            Object {
              "prim": "operation",
            },
          ],
          "prim": "NIL",
        },
        Object {
          "prim": "PAIR",
        },
      ],
    ],
    "prim": "code",
  },
]
`;
